clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -271,9 +289,9 @@ public void poll(long timeout) throws IOException {
                 try {
                     /* complete any connections that have finished their handshake */
                     if (key.isConnectable()) {
-                        channel.finishConnect();
-                        this.connected.add(channel.id());
-                        this.sensors.connectionCreated.record();
+                    	// queue this channel for completion after this loop
+                    	connectableChannels.add(channel);
+                    	continue;
```

Comment:
Yes, that may work. It just that it would be good to avoid duplicating the same code in two places.